src/main/kotlin/com/sourcegraph/cody/agent/CodyAgentService.kt

```diff
@@ -108,27 +115,35 @@ class CodyAgentService(project: Project) : Disposable {
     }
 
     @JvmStatic
-    fun applyAgentOnBackgroundThread(project: Project, block: Consumer<CodyAgent>) {
-      getInstance(project).getInitializedAgent(project).thenAccept { agent ->
-        ApplicationManager.getApplication().executeOnPooledThread { block.accept(agent) }
+    fun withAgent(project: Project): CompletableFuture<CodyAgent> {
+      val result = CompletableFuture<CodyAgent>()
+      if (CodyApplicationSettings.instance.isCodyEnabled) {
+        ApplicationManager.getApplication().executeOnPooledThread {
+          getInstance(project).getInitializedAgent(project, restartIfNeeded = false).thenAccept {
+            result.complete(it)
+          }
+        }
       }
+      return result
     }
 
     @JvmStatic
-    fun withAgentOnUIThread(project: Project, block: Consumer<CodyAgent>) {
-      getInstance(project).getInitializedAgent(project).thenAccept { agent ->
-        ApplicationManager.getApplication().invokeLater { block.accept(agent) }
+    fun withAgentRestartIfNeeded(project: Project): CompletableFuture<CodyAgent> {
+      val result = CompletableFuture<CodyAgent>()
+      if (CodyApplicationSettings.instance.isCodyEnabled) {
```

Comment:
What about the "else" case? 🤔 is it ok to NOT complete it? my thinking is it will make all the calls to the result waiting for a future that will never be completed.

Reply:
Hm, you are right, as previously it was a bit different (callback was never called).
Still, there ideally should be no calls to agent when cody is disabled, so this is mostly additional safeguard. 
Let me think for a moment if we can do anything more useful there.